parsing/cLexer.ml

```diff
@@ -773,18 +773,12 @@ let rec next_token ~diff_mode loc s =
 (** {6 The lexer of Coq} *)
 
 let func next_token ?(loc=Loc.(initial ToplevelInput)) cs =
-  let bp_ = Loc.(loc.bp) in
   let cur_loc = ref loc in
   Gramlib.LStream.from ~loc
     (fun () ->
       let (tok, loc) = next_token !cur_loc cs in
       cur_loc := after loc;
-      let aloc = Loc.{loc with bol_pos = loc.bol_pos + bp_;
-                                    bp = loc.bp + bp_;
-                                    ep = loc.ep + bp_} in
-     (* Debug: uncomment this for tracing tokens seen by coq...*)
-     (*  Printf.eprintf "(line %i, %i-%i)[%s]\n%!" aloc.line_nb aloc.bp aloc.ep (Tok.extract_string diff_mode t);*)
-      Some (tok,aloc))
```

Comment:
The debugger needs this code to get correct offsets for breakpoints.  Please figure out a way to support it.  Also, please test it in the debugger.

Reply:
How do you ensure that the returned `bol_pos` is correct?

Reply:
`Stream.count` is now set correctly at `idetop.ml` , which is what determines `bp`, so TTBOMK this should be correct, as described in the commit message.

Reply:
> How do you ensure that the returned `bol_pos` is correct?

More concretely, CLexer will update `bol_pos` from `Stream.count`, so indeed maybe you want to pass `bol_pos` to the initial `loc` in `idetop.ml` ? I didn't do so as before it wasn't done, but I've pushed a commit to do so.

Anyways, these kind of stuff should be covered by fake_ide tests, as I've done with the unit test here.

Reply:
Patch tested, seems to work fine!

Reply:
Last patch doesn't make a different as far as I can see tho.

Reply:
I won't have time to fully dig into this (or try it) until Thursday or Friday.  Does this have to be in 8.17?

I earlier made some performance fixes due to side effects of the debugger changes.  In its current state, I think this PR will break the positioning of tooltips in the script buffer (not breakpoint offsets as I said earlier).

The `let bol_pos = bp - bol_byte` in `coqOps.ml` (from #15787) passes `bol_pos` _relative_ to `bp`, which works with the existing code because `bol_pos = loc.bol_pos + bp_` on line 782 above adds `bp` back to return an _absolute_ value.  So you may need to change that line to just `let bol_pos = bol_byte`.  This assumes that vsCoq and coqtail are simply passing 0 for bp, which is probably true.  But it a tweak to the semantics of the message; a more conservative approach would be to add bp to bol_pos on the Coq side when you initialize the stream.

If `bol_pos` is incorrect, IIRC tooltips for warnings will be applied in the wrong spot in the script buffer.  A reasonable test would be to run a script that generates a warning or two and verify that tooltips are attached to the correct part of the script buffer.  Note that the sentence that generates the warning must not be the first sentence in the script (that wouldn't have the error).

Reply:
You may need to generate the warning on the second sentence in a single line to see the error if there is one.

Take my analysis with a grain of salt, I didn't have time to run the code and verify it.

Reply:
@jfehrle did you get a chance to look at Emilio's answers?

Reply:
I've been out of town for several days.  I will look at this later today.  I'd like to understand what broke it in master and to verify that it doesn't introduce a performance issue.

Reply:
The last commit ("Fix bizarre location-handling code.") reintroduces the O(n^2) behavior.  The ones before it are OK. 
(Pff.v, a 1.2 MB file, takes about 3 minutes to run in CoqIDE both on master and with the first 3 commits.  With the 4th commit, I gave up after 10 minutes, and it had only processed half the source.)

If the last commit is omitted, I think this change is OK.  The change I suggest [here](https://github.com/coq/coq/pull/16978/files#r1045382094) is not critical.

Reply:
Thanks!

Reply:
Last commit indeed was bogus, removed from the set.